### ISSUE TITLE
FakeAsserter.AssertWasCalled is not thread-safe

### ISSUE THREAD

**Issue by tillig** [FakeItEasy/FakeItEasy#600]
I have some fakes set up that raise events which run a handler on a separate thread. My test executes and verifies that the handler has executed and done its work. This async execution is causing `MustHaveHappened()` assertions to fail.

Probably best explained with an example...

I have a _cache_ and a _cache proxy_. Think distributed cache with a proxy wrapped around it holding a copy of things in memory. When someone _writes_ to the cache proxy, it updates its in-memory set of information and then sends a request to the distributed cache to update its remote info. When the distributed cache has any updates, it notifies the cache proxy that it should refresh its local data from the distributed cache.

It does end up doing a bit of circular-ish logic like:

1. User writes to the cache proxy.
2. Cache proxy writes to local storage.
3. Cache proxy writes to distributed cache.
4. Distributed cache gets new value and fires change event.
5. Cache proxy receives change event (on a separate/background thread) and refreshes the local contents.

The fake is set up like this:

``` csharp
var cache = A.Fake<ICache>();
var proxy = new CacheProxy(cache);
A.CallTo(() => cache.Set(A<Key>.Ignored, A<Value>.Ignored))
  .Invokes(call => {
    var thread = new Thread(() => proxy.Reload());
    thread.IsBackground = true;
    thread.Start();
  });
```

This actually executes perfectly - the cache proxy "Set" method calls the cache "Set" method and on a background thread the proxy gets a signal to Reload.

Depending on the threading timing, I get an exception like this:

```
System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at System.Collections.Generic.List`1.Enumerator.MoveNext()
   at System.Linq.Enumerable.Count[TSource](IEnumerable`1 source, Func`2 predicate)
   at FakeItEasy.Core.FakeAsserter.AssertWasCalled(Func`2 callPredicate, String callDescription, Func`2 repeatPredicate, String repeatDescription)
   at FakeItEasy.Configuration.RuleBuilder.MustHaveHappened(Repeated repeatConstraint)
   at FakeItEasy.Configuration.RuleBuilder.ReturnValueConfiguration`1.MustHaveHappened(Repeated repeatConstraint)
   at FakeItEasy.AssertConfigurationExtensions.MustHaveHappened(IAssertConfiguration configuration)
   at MyTests.CacheProxyFixture.Set_UpdatesValue()

Collection was modified; enumeration operation may not execute.
```

I think that's happening because [FakeAsserter.cs line 28](https://github.com/FakeItEasy/FakeItEasy/blob/aa467875f94a493ba3dc0d9d0a9c98d0df2ca979/src%2FFakeItEasy%2FCore%2FFakeAsserter.cs#L28) uses the LINQ `Count()` method but the background thread has modified the list of calls in the meantime.

**It might be good to do a `.ToList()` or something similar during assertions** to avoid that.

While I recognize that my test would still technically fail (I'll still have to figure that out), it'd fail due to the call not having been made by the time I called the assert rather than a hard-to-troubleshoot LINQ error.

**Comment by blairconrad** [comment]
@tillig thanks for raising this. My understanding is that you have an `Invokes` that starts a background thread and then later you call `MustHaveHappened` without regard to whether background thread has completed?

Excellent point about how the way in which FakeItEasy is enumerating the list competes with the modifications to it, but

1. You're right that this test will fail anyhow, and that you'd be much better off waiting for the thread to be done before asserting. 
2. To add an additional wrinkle, I am not sure that `ToList` wouldn't run afoul of the same problem that `Count` does. Are you? I'd expect that we'd iterate over the original list and still may be caught enumerating while the list is being modified.

While I like to give as informative an error messages as possible, I'm not sure what the path forward is here. Also, we've never claimed that individual Fake objects will be threadsafe, nor am I sure making them so is worthwhile.

Since you'll have to modify your test to make it consistently pass (or fail) anyhow, I'd look into that. I think it will resolve your problem with the list.

**Comment by blairconrad** [comment]
When I try to `ToList` a list that's having elements added, I am getting an error (not every time of course, but sometimes):

```
System.ArgumentException : Destination array was not long enough. Check destIndex and length, and the array's lower bounds.
at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length, Boolean reliable)
at System.Collections.Generic.List`1.CopyTo(T[] array, Int32 arrayIndex)
at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
at System.Linq.Enumerable.ToList(IEnumerable`1 source)
at ThreadedToList.Class1.MethodName_StateUnderTest_ExpectedBehavior() in Class1.cs: line 24 
```

So I think the `ToList` is a dead end. We could use a concurrent collection of some sort, but given the usual/recommended uses, I don't think it's worthwhile.

**Comment by tillig** [comment]
Hmmm. It does seem pretty edge-case, to be sure. I wasn't sure about the `ToList` myself, but it seems you proved out that won't work. :frowning:

I don't see that use of thread-safe collections in more places would hurt much, though. It might be slightly less performant, but I'm not sure mocking frameworks really have performance constraints so much.

I was able to fix my test, so I'm not blocked, but I did see a couple of other issues out there involving thread safety like #60, #484, so I don't think threading things are just seen by me. Might be a general feature to look into for the 2.0 release?

**Comment by thomaslevesque** [comment]
> I don't see that use of thread-safe collections in more places would hurt much, though

The problem with concurrent collections is that they're only available since .NET 4.0, and we still support 3.5...

**Comment by tillig** [comment]
Given [MS end of life for .NET 3.5SP1 was January of last year](http://stackoverflow.com/questions/2713392/when-will-microsoft-end-mainstream-support-for-existing-versions-of-the-net-fra) I'd like to think (though I suppose I could be wrong) that such a constraint wouldn't be on the 2.0 release of FakeItEasy.

**Comment by blairconrad** [comment]
> The problem with concurrent collections is that they're only available since .NET 4.0, and we still support 3.5...

We could cobble something together, @thomaslevesque. We do have shims for some concurrent collections already. But really it would more likely be about how we used the collections than their inherent threadsafety, and tracking down the changes we'd need to make to say "Fakes are threadsafe, without compromising the workflows people are using them for" could be quite a lot of work.

> Given MS end of life for .NET 3.5SP1 was January of last year I'd like to think (though I suppose I could be wrong) that such a constraint wouldn't be on the 2.0 release of FakeItEasy.

I'm probably misunderstanding everything, but from [Microsoft Support Lifecycle FAQ](https://support.microsoft.com/en-us/lifecycle#gp/Framework_FAQ):
> .NET Framework 3.5 SP1: Beginning with .NET Framework 3.5 Service Pack 1 (SP1) the .NET Framework is defined as a component instead of an independent product. So, support for .NET Framework 3.5 SP1 is driven by the support lifecycle policy of the Windows operating system (i.e. when Windows is in the Mainstream Support phase or Extended Support phase, so are its components). When a product such as Windows reaches the end of support, so do its related components. .NET Framework 3.5 SP1 is supported on Windows Server 2003 SP2, Windows Vista SP2, Windows 7 SP1, Windows Server 2008 SP2, Windows Server 2008 R2 SP1, Windows 8.1 Update, Windows Server 2012, Windows Server 2012 R2, and Windows 10 according to the end of support date for each operating system.

And according to [Windows lifecycle fact sheet](http://windows.microsoft.com/en-ca/windows/lifecycle), Windows 10 is supported until 
October 13, 2020 (mainstream) or October 14, 2025 (extended).

However, that doesn't mean that FakeItEasy has to support 3.5. We could drop it at any time and allow people to use older versions of FakeItEasy. We've discussed it, in fact, and someday we'll may decide to do so, should the cost of keeping 3.5 support prove too high. Or we could keep supporting it forever, I suppose. In the meantime, we'd like to keep it going as long as feasible.

> I was able to fix my test, so I'm not blocked, but I did see a couple of other issues out there involving thread safety like #60, #484, so I don't think threading things are just seen by me. Might be a general feature to look into for the 2.0 release?

@tillig, #60 has become a general catch-all "threading" issue because people keep going to it when any concurrency problem appears to rear its ugly head. There's no known specific actionable complaint in there at this point, as once we've identified what's causing people's problems, we've spun off specific issues.

#484 is a specific complaint about the FakeScope class, and that class _should_ be able to handle multiple threads. I'm glad you brought it up, since the issue should've been marked as in-progress and part of the 2.0 milestone, as we have a community member working on it. (Although there's a small technical roadblock.)

All that aside, I'm glad you fixed your test. Did you have to do anything specifically to work around the Fake's threadunsafety, or did fixing the problems related to your test logic make things work? If the former, then that adds weight to the idea that we might want to look into the situation you encountered. Otherwise, no _added_ weight.

> I don't see that use of thread-safe collections in more places would hurt much, though. It might be slightly less performant, but I'm not sure mocking frameworks really have performance constraints so much.

I see your point. The performance would probably be degraded a little. How much, I don't know. Nor do we know whether it would be a big deal—we've had complaints about lack of threadsafety and complaints about performance as well. Depending on what people are doing with the library, they may be affected by a slowdown and not by threadunsafety. (Maybe someone is using Fakes in a single thread, but for some reason making many thousands of calls.) It's very difficult to know which use case is the one that matters more. The problem is that by chasing threadsafety we'd be considering potentially a large amount of work in order to cover some problems that people may not even be encountering.

Then again, I always look forward to being shown when I'm being short-sighted. Convince me!

**Comment by tillig** [comment]
It may be good to just file this under dupe of #60 (the concurrency catch-all) and call it a day.

In order to work around the thread safety problem, I had to make my event less "fire and forget" and more "fire and await a ManualResetEvent for one second that the event handler may or may not signal depending on the test failure conditions." I'm not sure if you'd classify that as working around the thread unsafety or fixing the test... but it's probably a little of both.

I'm not sure you'd have to go "all in" to declare ultimate thread safety across everything. It could be more incremental - fix things as you find them, like you're doing now. I agree it could be a lot of work to try to get it 100% right.

I've never run into anything I'd consider a perf problem with any unit testing or mocking framework. I'd like to think that unit tests themselves are pretty fast and one or two milliseconds isn't going to make or break anyone. Granted, if it's 250ms _per test_, that's something that can add up, but I don't think concurrent collections or appropriately used locking will be turning up the heat like that on a single test. Maybe _across a whole test run_, which is negligible at worst.

I will say this: I've used several mocking frameworks, from Typemock to Moq to Rhino. Each has their problems... but I can't ever say I had a thread safety issue in any of the others. When they failed, I could always nail down exactly what in my test was causing it - no mysterious "enumeration was changed" errors. I don't know if that adds weight to the case or not, but it might be a "feature parity" sort of item to consider.

I mean, **at this point I'd be happy if the `InvalidOperationException` was caught on the `.Count()` call and a more meaningful message was rethrown.** Changing the question/request - **it may be good enough to just provide a better message** to help someone like me troubleshoot the test and work around it. I think the largest frustration I had with the issue was that I had to track down _what_ was going wrong and _why_ inside _the mocking framework_ instead of getting down to writing product code.

Which isn't at all to say that I don't value the great work you guys are doing here. **Love the syntax, love the product - it's why we switched!** Just to say that when, say, your test fails in XUnit because of an exception _inside XUnit_ and not in your test, and there's nothing you can do but dive into XUnit's source and debug _the unit testing framework_... well, that's kinda painful, you know?

**Comment by adamralph** [comment]
> It may be good to just file this under dupe of #60 (the concurrency catch-all) and call it a day.

#60 is already closed in favour of more specific issues.

**Comment by blairconrad** [comment]
@tillig said:

> In order to work around the thread safety problem, I had to make my event less "fire and forget" and more "fire and await a ManualResetEvent for one second that the event handler may or may not signal depending on the test failure conditions." I'm not sure if you'd classify that as working around the thread unsafety or fixing the test... but it's probably a little of both.

To tell the truth, it sounds like it's fixing the test. Had you not done that, you wouldn't've been able to guarantee that you were waiting long enough to check that the call must've happened, and you would've been faced with a sometimes passing/sometimes failing test. (As you point out, a test using NSubstitute, which seems to work a little more toward making their substitutes threadsafe, exhibits exactly this behaviour.)
Whether that's preferable to having a test throw a concurrent modification exception I think is a little in the eye of the beholder.

Me, were I writing tests that explicitly created threads or that tested multi-threaded code, and I saw a concurrent modification exception, the very first thing that I would think would be that there was a problem due to the multithreaded nature of the code that I was writing, and then I think I'd pursue making sure I had a valid test, without regard to what the framework was doing.

Then again, it's easy for me to say that, since I've been indoctrinated not to use fakes (or many other objects!) in a threadunsafe manner. So, you already primed me by explaining that there was thread unsafety, so I had it in my mind. On a different day, maybe I wouldn't've thought of that.

Also, it's not like we're _guaranteeing_ that we'd throw that exception, so relying on it is not foolproof.

All that being said, I'm coming around. There's certainly no harm in an error message that would better convey what went wrong.
And for this particular collection, there's probably no advantage to allowing it to be modified by multiple threads concurrently. Quickly eyeballing it, it could probably be replaced with a ConcurrentStack (or shim for 3.5). The ConcurrentStack should be fast enough.

@adamralph, @thomaslevesque, do you have any preference? I think we have an **enhancement** here, whether it's the message or making `FakeManager.recordedCallsField` a little more proof against multithreading. I'm still not keen on "a general [multithreading] feature to look into for the 2.0 release", though.

**Comment by blairconrad** [comment]
Possibly `ConcurrentQueue`. I have some sort of cold and can't brain.

**Comment by adamralph** [comment]
Can we get a TL;DR?

If it's another feature relying on ambient state, then we probably have to think about making it instance based instead, a la #602.

**Comment by blairconrad** [comment]
@adamralph, your wish is my command.

**TL;DR:**

This is **not** another feature relying on ambient state. It's already instance based. @tillig was using a Fake from multiple threads concurrently. If we want, we can shut down the issue right there, as this is known not to be safe.

The list of recorded calls, `FakeManager.recordedCallsField` (accessed by `FakeManager.AllRecordedCalls`), had elements added to it at the same time that `FakeAsserter.AssertWasCalled` was enumerating the list, to verify that a call was made.

I pointed out that this was not a sanctioned use, and that even if there were no exception, the test would not pass reliably, as there was a race condition between action and assertion. @tillig amended his test, fixing the error, and all was well. But he persisted in his stance that the exception thrown from `FakeAsserter` was confusing, and that making `recordedCallsField` proof against multiple threads would be an advantage. Or at least trapping the error and providing "a more meaningful message".

I didn't really see any harm, so polled the group to see if you all preferred one approach or another.

**Comment by tillig** [comment]
I'd even go for some docs on the doc site - on the events page, on the async methods page, in the troubleshooting section - _something_ talking about the officially supported (or not supported) thread safety and places you might see this crop up.

**Comment by adamralph** [comment]
@blairconrad thanks very much, that's a very useful précis.

Switching to a concurrent collection type sounds reasonable, and we could even write a shim for .NET 3.5 similar to [ConcurrentDictionary](https://github.com/FakeItEasy/FakeItEasy/blob/14f149d21b1ecd2f13e06c0cec7a94b3f7682e37/src/FakeItEasy.Net35/Compatibility/ConcurrentDictionary.cs).

**Comment by blairconrad** [comment]
Lovely. I'll up-for-grabs it.

Unless you'd like it, @tillig? It seems to be a cause near and dear to your heart!

**Comment by tillig** [comment]
I can take a look at it, sure.

**Comment by blairconrad** [comment]
Thanks. You're a sport!

**Comment by blairconrad** [comment]
Thanks for the merge, folks.

@tillig, thanks for your work on this.
Look for your name in the release notes. :trophy:

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).
